--- vcfnorm.c	2018-08-26 12:00:00.000000000 -0400
+++ vcfnorm.c	2018-08-26 12:00:00.000000000 -0400
@@ -96,7 +96,7 @@
     char **argv, *output_fname, *ref_fname, *vcf_fname, *region, *targets;
     int argc, rmdup, output_type, n_threads, check_ref, strict_filter, do_indels;
     int nchanged, nskipped, nsplit, ntotal, mrows_op, mrows_collapse, parsimonious;
-    int record_cmd_line;
+    int record_cmd_line, keep_sum_ad;
 }
 args_t;
 
@@ -617,7 +617,7 @@
         assert( nvals>0 ); \
         type_t *vals = (type_t *) args->tmp_arr1; \
         int len = bcf_hdr_id2length(args->hdr,BCF_HL_FMT,fmt->id); \
-        int i, nsmpl = bcf_hdr_nsamples(args->hdr); \
+        int i, j, nsmpl = bcf_hdr_nsamples(args->hdr); \
         if ( nvals==nsmpl ) /* all values are missing */ \
         { \
             bcf_update_format_##type(args->hdr,dst,tag,vals,nsmpl); \
@@ -648,6 +648,8 @@
             for (i=0; i<nsmpl; i++) \
             { \
                 dst_vals[0] = src_vals[0]; \
+                if (args->keep_sum_ad && fmt->id==bcf_hdr_id2int(args->hdr,BCF_DT_ID,"AD") ) \
+                  for (j=1; j<nvals; j++) if ( j!=ialt+1 ) dst_vals[0] += src_vals[j]; \
                 dst_vals[1] = src_vals[ialt+1]; \
                 dst_vals += 2; \
                 src_vals += nvals; \
@@ -1850,6 +1852,7 @@
     fprintf(stderr, "    -d, --rm-dup <type>               remove duplicate snps|indels|both|all|none\n");
     fprintf(stderr, "    -f, --fasta-ref <file>            reference sequence (MANDATORY)\n");
     fprintf(stderr, "    -m, --multiallelics <-|+>[type]   split multiallelics (-) or join biallelics (+), type: snps|indels|both|any [both]\n");
+    fprintf(stderr, "    -k, --keep-sum-AD                 keep sum AD vector constant\n");
     fprintf(stderr, "        --no-version                  do not append version and command line to the header\n");
     fprintf(stderr, "    -N, --do-not-normalize            do not normalize indels (with -m or -c s)\n");
     fprintf(stderr, "    -o, --output <file>               write output to a file [standard output]\n");
@@ -1875,6 +1878,7 @@
     args->output_type = FT_VCF;
     args->n_threads = 0;
     args->record_cmd_line = 1;
+    args->keep_sum_ad = 0;
     args->aln_win = 100;
     args->buf_win = 1000;
     args->mrows_collapse = COLLAPSE_BOTH;
@@ -1888,6 +1892,7 @@
         {"fasta-ref",required_argument,NULL,'f'},
         {"do-not-normalize",no_argument,NULL,'N'},
         {"multiallelics",required_argument,NULL,'m'},
+        {"keep-sum-AD",no_argument,NULL,'k'},
         {"regions",required_argument,NULL,'r'},
         {"regions-file",required_argument,NULL,'R'},
         {"targets",required_argument,NULL,'t'},
@@ -1904,7 +1909,7 @@
         {NULL,0,NULL,0}
     };
     char *tmp;
-    while ((c = getopt_long(argc, argv, "hr:R:f:w:Dd:o:O:c:m:t:T:sN",loptions,NULL)) >= 0) {
+    while ((c = getopt_long(argc, argv, "hr:R:f:w:Dd:o:O:c:m:t:T:sNk",loptions,NULL)) >= 0) {
         switch (c) {
             case 'N': args->do_indels = 0; break;
             case 'd':
@@ -1929,6 +1934,7 @@
                     else error("The argument to -m not recognised: %s\n", optarg);
                 }
                 break;
+            case 'k': args->keep_sum_ad = 1; break;
             case 'c':
                 if ( strchr(optarg,'w') ) args->check_ref |= CHECK_REF_WARN;
                 if ( strchr(optarg,'x') ) args->check_ref |= CHECK_REF_SKIP;
